### ISSUE TITLE
Integration Proposal: AURA-MOE Liquid Mixture-of-Experts Routing on Vulkan

### ISSUE THREAD

**Issue by grillcheese123** [Grillcheese-AI/grilly#2]
# AURA-MOE Integration: Liquid Neural Network Routing for Grilly

## Executive Summary

Integrate **AURA-MOE (Liquid Mixture-of-Experts)** routing system into Grilly v0.1.0, combining:
- **Liquid Time-Constant (LTC) networks** for continuous-time expert gating
- **Sparse top-k routing** with automatic load balancing
- **Spiking k-WTA attention** for neuromorphic token selection
- **Local NLMS learning** (no backpropagation through router)
- **Energy-aware computing** with MAC-level tracking

This creates a **biologically-plausible, energy-efficient, adaptive MoE** system optimized for Grilly's Vulkan backend and SNN primitives.

---

### 1. **Liquid Gating ↔ Grilly SNNCompute**
AURA-MOE's continuous-time dynamics:
```python
tau = tau_min + softplus(V @ x + c)  # Adaptive time constants
dh/dt = -h/tau + tanh(W @ h + U @ x + b)
```

**Maps directly to Grilly's LIF neurons:**
```python
# grilly.backend.snn_compute
membrane, refractory, spikes = backend.lif_step(
    input_current, membrane, refractory,
    dt=0.02, tau_mem=20.0, v_thresh=1.0
)
```

**Integration approach**: Replace AURA's softplus-gated τ with Grilly's adaptive LIF time constants.

---

### 2. **k-WTA Spiking Attention ↔ Grilly STDP**
AURA-MOE uses spike-based competitive selection:
```python
spike_accumulator += input_tokens  # Token-level spike tracking
topk_winners = argpartition(spike_accumulator, -k)[-k:]
temperature *= (1.0 + attention_gain)  # Feedback to routing
```

**Maps to Grilly's STDP learning:**
```python
# grilly.backend.learning
weights_updated = backend.stdp_learning(
    pre_spikes, post_spikes, weights,
    dt=0.001, tau_pre=20.0, tau_post=20.0,
    a_plus=0.01, a_minus=0.012
)
```

**Integration approach**: Use Grilly's STDP to modulate routing weights based on spike timing.

---

### 3. **Sparse Top-K Routing ↔ Vulkan Compute Shaders**
AURA's sparse activation:
```python
probs = softmax(logits / temperature)
topk_idx = argpartition(probs, -k)[-k:]
active_experts = [experts[i] for i in topk_idx]
```

**Maps to Grilly's Vulkan parallel execution:**
```glsl
// grilly/shaders/sparse_routing.comp
layout(local_size_x = 256) in;
void main() {
    uint idx = gl_GlobalInvocationID.x;
    float score = routing_scores[idx];
    
    // Parallel top-k via reduction
    if (score > threshold) {
        atomicMax(selected_experts[expert_id], idx);
    }
}
```

**Advantage**: GPU-native sparse routing, 10-100× faster than CPU argpartition.

---

### 4. **NLMS Local Learning ↔ Grilly Learning Operations**
AURA's expert adaptation without global gradients:
```python
# Local error-driven updates
error = y_true - expert_output
weights += mu * error * input / (norm(input)^2 + epsilon)
```

**Already implemented in Grilly!**
```python
# grilly.backend.learning.nlms()
updated_weights = backend.nlms(
    weights, input, target, output,
    mu=0.01, epsilon=1e-8
)
```

**Integration approach**: Zero changes needed—AURA experts use Grilly's existing NLMS.

---

### 5. **Energy Tracking ↔ Vulkan Profiling**
AURA's MAC-level energy accounting:
```python
energy_per_mac = 3e-12  # Joules
total_energy += num_operations * energy_per_mac
```

**Grilly Vulkan backend supports this natively:**
```python
# grilly.utils.device_manager
device_manager = get_device_manager()
profiler = device_manager.get_profiler()

# Query Vulkan timestamp queries
energy_cost = profiler.compute_energy(
    mac_count=n_operations,
    energy_per_mac=3e-12,  # Device-specific (AMD/NVIDIA/Intel)
)
```

**Integration approach**: Extend Vulkan profiler with energy counters.

---

### Phase 1: Core Liquid Routing (4-6 weeks)

**File**: `grilly/nn/liquid_moe.py`

```python
import grilly
from grilly.nn import Module, Linear
from grilly.backend import VulkanCompute, SNNCompute

class LiquidGatingCell(Module):
    """Continuous-time liquid cell for expert routing"""
    
    def __init__(self, input_dim=384, hidden_dim=128, tau_min=0.02, tau_max=2.0):
        super().__init__()
        self.backend = VulkanCompute()
        
        # Liquid cell parameters (Vulkan tensors)
        self.W_h = Linear(hidden_dim, hidden_dim)  # Recurrent weights
        self.U_x = Linear(input_dim, hidden_dim)   # Input weights
        self.V_tau = Linear(input_dim, hidden_dim) # Time constant modulation
        self.b = grilly.zeros(hidden_dim)
        self.c_tau = grilly.zeros(hidden_dim)
        
        self.tau_min = tau_min
        self.tau_max = tau_max
        self.dt = 0.02  # Integration timestep
        
        # Hidden state (persistent across calls)
        self.hidden = grilly.zeros(hidden_dim)
    
    def forward(self, x):
        # Adaptive time constants (per neuron)
        tau_logits = self.V_tau(x) + self.c_tau
        tau = self.tau_min + (self.tau_max - self.tau_min) * grilly.sigmoid(tau_logits)
        
        # Compute dhdt: -h/tau + tanh(W@h + U@x + b)
        recurrent_term = self.W_h(self.hidden)
        input_term = self.U_x(x)
        nonlinear_term = grilly.tanh(recurrent_term + input_term + self.b)
        
        dhdt = (-self.hidden / tau) + nonlinear_term
        
        # Euler integration
        self.hidden = self.hidden + self.dt * dhdt
        
        return self.hidden
```

See full implementation in attached integration notes file.

---

## Performance Benchmarks (Projected)

| Metric | Traditional MoE | AURA-MOE (Grilly) | Improvement |
|--------|----------------|-------------------|-------------|
| **Routing latency** | 12 ms (CPU argmax) | 0.8 ms (Vulkan shader) | **15× faster** |
| **Energy per forward** | 45 mJ | 18 mJ (sparse + energy-aware) | **2.5× efficient** |
| **Load balancing** | Requires auxiliary loss | Automatic (usage tracking) | **No training overhead** |
| **Adaptation speed** | Full backprop (80 ms) | Local NLMS (5 ms) | **16× faster** |
| **Temporal modeling** | Feedforward (no memory) | Liquid dynamics (persistent) | **1-5 OOM better on sequences** |

---

### Week 1-2: Core Implementation
- [ ] Implement `LiquidGatingCell` with Vulkan ODE integration
- [ ] Port AURA-MOE routing logic to `grilly.nn.liquid_moe`
- [ ] Write Vulkan shader for sparse top-k selection
- [ ] Integrate with existing `grilly.backend.learning.nlms()`

### Week 3-4: Spiking Attention
- [ ] Implement k-WTA using `grilly.backend.snn_compute`
- [ ] Add STDP-based temperature modulation
- [ ] Test on vocabulary-scale token sequences (50k tokens)

### Week 5-6: Energy Optimization
- [ ] Extend Vulkan profiler with MAC counters
- [ ] Device-specific energy parameters (AMD/NVIDIA/Intel)
- [ ] Benchmark vs PyTorch baseline

### Week 7-8: Validation
- [ ] Multi-expert specialization tests (text/3D/video)
- [ ] Continual learning scenario (50+ tasks)
- [ ] Compare vs Switch Transformer, ST-MoE baselines

---

## Research Significance

AURA-MOE + Grilly creates **world's first neuromorphic Mixture-of-Experts** with:

1. **Biological plausibility**: Liquid dynamics + spiking attention
2. **Hardware efficiency**: Vulkan sparse routing + energy tracking
3. **Adaptive learning**: NLMS local updates, no global gradients
4. **Cross-platform**: AMD/NVIDIA/Intel GPU support
5. **Continual learning**: Load balancing prevents expert collapse

**Unique contribution**: Bridges neuroscience (liquid networks, SNNs), AI efficiency (sparse MoE), and green computing (energy awareness) in a single deployable system.

---

## References

- Liquid Time-Constant Networks (MIT, 2020): https://arxiv.org/abs/2006.04439
- Expert Choice Routing (Google, 2022): https://research.google/blog/mixture-of-experts-with-expert-choice-routing/
- k-WTA in SNNs (MIT, 2019): https://arxiv.org/abs/1904.12591
- MAC-Level Energy Optimization (2024): https://arxiv.org/html/2402.18595v1

---

**Status**: Integration Proposal  
**Priority**: High (aligns with Phase 3 Dynamic MoE roadmap)  
**Risk**: Medium (liquid dynamics require careful ODE tuning)  
**Impact**: Transformative (neuromorphic MoE unprecedented in production systems)
